test/internet/test-http-host-header-ipv6-fail.js

```diff
@@ -0,0 +1,51 @@
+'use strict';
+/*
+ * When using the object form of http.request and using an IPv6 address
+ * as a hostname, and using a non-standard port, the Host header
+ * is improperly formatted.
+ * Issue: https://github.com/nodejs/node/issues/5308
+ * As per https://tools.ietf.org/html/rfc7230#section-5.4 and
+ * https://tools.ietf.org/html/rfc3986#section-3.2.2
+ * the IPv6 address should be enclosed in square brackets
+ */
+
+const common = require('../common');
+const assert = require('assert');
+const http = require('http');
+
+var hadError = 0;
+var hadRequest = false;
+var server = null;
+const hostname = '::1';
+const port = common.PORT;
+
+function httpreq() {
+  var req = http.request({
+    host: hostname,
+    port: port,
+    path: '/',
+    method: 'GET'
+  });
+  req.end();
+}
+
+server = http.createServer(function(req, res) {
```

Comment:
if you wrap this in a `common.mustCall(function(req, res) { })`, then the `hadRequest` can be be totally removed. It doesn't look like `hadError` is used at all, so that can be removed. With those two gone, we can also remove the `process.on('exit')` listener.